packages/langium/src/grammar/type-system/type-validator.ts

```diff
@@ -193,18 +193,18 @@ function checkPropertiesConsistency(inferred: Property[], declared: Property[],
             }
 
             if (checkOptional(foundProperty, expectedProperty) && !expectedProperty.optional && foundProperty.optional) {
-                errorToAssignment(foundProperty.name, `A property '${foundProperty.name}' can't be optional.`);
+                errorToRuleNodes(`A property '${foundProperty.name}' can't be optional`);
             }
         } else {
             errorToAssignment(foundProperty.name, `A property '${foundProperty.name}' is not expected.`);
         }
     }
 
     // detects lack of properties
-    for (const foundProperty of declared) {
-        const expectedProperty = inferred.find(e => foundProperty.name === e.name);
-        if (!expectedProperty) {
-            errorToRuleNodes(`A property '${foundProperty.name}' is expected`);
+    for (const expectedProperty of declared) {
```

Comment:
Just a note, I fixed these namings as they appeared to be inconsistent. Expected should be declared, and inferred being found, matching what's done above in the same function.